python_modules/dagster/dagster/_core/instance/__init__.py

```diff
@@ -2541,7 +2541,7 @@ def report_run_canceled(
         return dagster_event
 
     def report_run_failed(
-        self, dagster_run: DagsterRun, message: Optional[str] = None
+        self, dagster_run: DagsterRun, message: Optional[str] = None, job_failure_data=None
```

Comment:
```job_failure_data: Optional[JobFailureData] = None```

Reply:
I actually tried this, but it didn't look like importing `JobFailureData` into `instance/__init__.py` was supported. I figured that was why a number of imports were inside the class methods instead of at the head of the file. 

So I wasn't sure how to make `JobFailureData` available for type hinting in the method declaration.

e.g., a change like 
```
@@ -47,6 +47,7 @@
     DagsterRunAlreadyExists,
     DagsterRunConflict,
 )
+from dagster._core.events import JobFailureData
 from dagster._core.log_manager import get_log_record_metadata
 from dagster._core.origin import JobPythonOrigin
 from dagster._core.storage.dagster_run import (
@@ -2541,7 +2542,7 @@ def report_run_canceled(
         return dagster_event

     def report_run_failed(
-        self, dagster_run: DagsterRun, message: Optional[str] = None, job_failure_data=None
+        self, dagster_run: DagsterRun, message: Optional[str] = None, job_failure_data: Optional[JobFailureData] = None
     ) -> "DagsterEvent":
         from dagster._core.events import DagsterEvent, DagsterEventType
```
throws 

```
File ".../dagster/python_modules/dagster/dagster/_core/instance/__init__.py", line 50, in <module>
    from dagster._core.events import JobFailureData
  File ".../dagster/python_modules/dagster/dagster/_core/events/__init__.py", line 43, in <module>
    from dagster._core.definitions.partition import PartitionsSubset
ImportError: cannot import name 'PartitionsSubset' from partially initialized module 'dagster._core.definitions.partition' (most likely due to a circular import) (.../dagster/python_modules/dagster/dagster/_core/definitions/partition.py)
```

Reply:
Ah OK this should work then:

```
self, dagster_run: DagsterRun, message: Optional[str] = None, job_failure_data: "JobFailureData" =None
```

(note JobFailureData in quotes)

and then up top you can import JobFailureData only within the `if TYPE_CHECKING` block: https://github.com/dagster-io/dagster/blob/master/python_modules/dagster/dagster/_core/instance/__init__.py#L102-L177

Reply:
That looks like it worked, thanks! And thanks for pointing me at the details for the testing.